images/handlers.go

```diff
@@ -179,14 +181,104 @@ func Dispatch(ctx context.Context, handler Handler, limiter *semaphore.Weighted,
 			if len(children) > 0 {
 				return Dispatch(ctx2, handler, limiter, children...)
 			}
-
 			return nil
 		})
 	}
+	return eg.Wait()
+}
 
+// DispatchWithOpts runs the provided handler for content specified by the
+// descriptors with concurrent and duplication suppressor control.
+//
+// If the handler decodes subresources, they will be visited, as well. A handler
+// can return 'ErrSkipDesc` to signal to the dispatcher to not traverse any
+// children. If any handler returns an error, the dispatch session will be
+// canceled.
+func DispatchWithOpts(ctx context.Context, handler Handler, opts []DispatchOpt, descs ...ocispec.Descriptor) error {
+	control := &dispatchControl{
+		concurrencyLimiter:    pkgsemaphore.NewNoop(),
+		duplicationSuppressor: pkgsingleflight.NewNoop(),
+	}
+
+	for _, opt := range opts {
+		if err := opt(control); err != nil {
+			return err
+		}
+	}
+	return dispatchWithControl(ctx, handler, control, descs...)
+}
+
+func dispatchWithControl(ctx context.Context, handler Handler, control *dispatchControl, descs ...ocispec.Descriptor) error {
+	eg, ctx2 := errgroup.WithContext(ctx)
+	for _, desc := range descs {
+		desc := desc
+
+		if err := control.concurrencyLimiter.Acquire(ctx, 1); err != nil {
+			return err
+		}
+
+		eg.Go(func() error {
+			desc := desc
```

Comment:
Was this for a memory aliasing concern?

Reply:
Good catch! I think I can remove it~ just follow the original pattern.